File: nine-language-version/lessons/11-error-handling/error_handling.hs
-- Lesson 11: Error Handling in Haskell - Use Maybe/Either, avoid exceptions

import Control.Exception (catch, SomeException)

-- Using Maybe for operations that might fail
safeDivide :: Double -> Double -> Maybe Double
safeDivide _ 0 = Nothing
safeDivide x y = Just (x / y)

-- Using Either for errors with information
divide :: Double -> Double -> Either String Double
divide _ 0 = Left "Cannot divide by zero"
divide x y = Right (x / y)

-- Chaining operations
compute :: Double -> Double -> Either String Double
compute x y = do
    result1 <- divide x y
    result2 <- divide result1 2
    return result2

main :: IO ()
main = do
    putStrLn "=== Error Handling in Haskell ===\n"

    putStrLn "1. Maybe type:"
    print $ safeDivide 10 2
    print $ safeDivide 10 0

    putStrLn "\n2. Either type:"
    print $ divide 10 2
    print $ divide 10 0

    putStrLn "\n3. Error chaining:"
    print $ compute 10 2
    print $ compute 10 0

    putStrLn "\n=== Haskell Error Handling ==="
    putStrLn "- Maybe for nullable results"
    putStrLn "- Either for errors with info"
    putStrLn "- Avoid exceptions, use types"

File: nine-language-version/lessons/11-error-handling/error_handling.pl
% Lesson 11: Error Handling in Prolog - catch/throw and failure

:- initialization(main).

% Safe division with catch
safe_divide(X, Y, Result) :-
    catch(
        Result is X / Y,
        error(evaluation_error(zero_divisor), _),
        (writeln('   Error: Division by zero'), fail)
    ).

% Throwing exceptions
divide(_, 0, _) :-
    throw(error(zero_divisor, 'Cannot divide by zero')).
divide(X, Y, Result) :-
    Result is X / Y.

% Using catch
compute(X) :-
    catch(
        divide(10, X, Result),
        error(zero_divisor, Msg),
        (format('   Caught: ~w~n', [Msg]), fail)
    ),
    format('   Result: ~w~n', [Result]).

main :-
    writeln('=== Error Handling in Prolog ===\n'),

    writeln('1. catch/throw:'),
    (safe_divide(10, 2, R) -> format('   divide(10, 2) = ~w~n', [R]) ; true),
    safe_divide(10, 0, _),

    writeln('\n2. Error handling:'),
    compute(2),
    compute(0),

    nl,
    writeln('=== Prolog Error Handling ==='),
    writeln('- catch/throw for exceptions'),
    writeln('- Logical failure'),
    writeln('- Error terms'),

    halt(0).

:- catch(main, E, (writeln(E), halt(1))).
